src/python/pants/base/cmd_line_spec_parser.py

```diff
@@ -48,20 +57,30 @@ def _normalize_spec_path(self, path: str) -> str:
       normalized = ''
     return normalized
 
-  def parse_address_spec(self, spec: str) -> AddressSpec:
-    """Parse the given spec into an `AddressSpec` object.
+  def parse_spec(self, spec: str) -> Union[AddressSpec, FilesystemSpec]:
+    """Parse the given spec into an `AddressSpec` or `FilesystemSpec` object.
 
     :raises: CmdLineSpecParser.BadSpecError if the address selector could not be parsed.
     """
-
     if spec.endswith('::'):
       spec_path = spec[:-len('::')]
-      return DescendantAddresses(self._normalize_spec_path(spec_path))
-    elif spec.endswith(':'):
+      return DescendantAddresses(directory=self._normalize_spec_path(spec_path))
+    if spec.endswith(':'):
       spec_path = spec[:-len(':')]
-      return SiblingAddresses(self._normalize_spec_path(spec_path))
-    else:
+      return SiblingAddresses(directory=self._normalize_spec_path(spec_path))
+    if ':' in spec:
       spec_parts = spec.rsplit(':', 1)
-      spec_path = self._normalize_spec_path(spec_parts[0]) 
-      name = spec_parts[1] if len(spec_parts) > 1 else os.path.basename(spec_path)
-      return SingleAddress(spec_path, name)
+      spec_path = self._normalize_spec_path(spec_parts[0])
+      return SingleAddress(directory=spec_path, name=spec_parts[1])
+    if '*' in spec:
+      return FilesystemSpec(glob=spec)
+    if PurePath(spec).suffix:
+      return FilesystemSpec(glob=self._normalize_spec_path(spec))
+    spec_path = self._normalize_spec_path(spec)
+    if Path(self._root_dir, spec_path).is_file():
+      return FilesystemSpec(glob=spec_path)
+    # Else we apply address shorthand, i.e. `src/python/pants/util` -> `src/python/pants/util:util`
+    # TODO: Figure out what this should look like if (once?) filesystem specs allow directories.
```

Comment:
I'm leaning towards a future change adding an option that determines whether a/b/c should mean the directory a/b/c or the target a/b/c:c. Because in a v2 world, in which goals for the most part can operate on files, we might want it to mean the directory (at least by default). But for now this is fine.